Firestore/core/test/firebase/firestore/testutil/testutil.h

```diff
@@ -51,7 +52,135 @@ namespace testutil {
+/**
+ * Creates a boolean FieldValue.
+ *
+ * @param bool_value A boolean value that disallows implicit conversions.
+ */
+template <typename T,
+          typename = typename std::enable_if<std::is_same<bool, T>{}>::type>
+inline model::FieldValue Value(T bool_value) {
+  return model::FieldValue::FromBoolean(bool_value);
+}
+
+// Overload that captures integer literals. Without this, int64_t and double
+// are equally applicable conversions.
+inline model::FieldValue Value(int value) {
+  return model::FieldValue::FromInteger(value);
+}
+
+inline model::FieldValue Value(int64_t value) {
+  return model::FieldValue::FromInteger(value);
+}
+
+inline model::FieldValue Value(double value) {
+  return model::FieldValue::FromDouble(value);
+}
+
+inline model::FieldValue Value(Timestamp value) {
+  return model::FieldValue::FromTimestamp(std::move(value));
+}
+
+inline model::FieldValue Value(const char* value) {
+  return model::FieldValue::FromString(value);
+}
+
+inline model::FieldValue Value(const std::string& value) {
+  return model::FieldValue::FromString(value);
+}
+
+inline model::FieldValue Value(const GeoPoint& value) {
+  return model::FieldValue::FromGeoPoint(value);
+}
+
+// This overload allows Object() to appear as a value (along with any explicitly
+// constructed FieldValues).
+inline model::FieldValue Value(const model::FieldValue& value) {
+  return value;
+}
+
+inline model::FieldValue ArrayValue(std::vector<model::FieldValue>&& value) {
+  return model::FieldValue::FromArray(std::move(value));
+}
+
+namespace details {
+
+/**
+ * Recursive base case for AddPairs, below. Returns the map.
+ */
+inline model::FieldValue::Map AddPairs(const model::FieldValue::Map& prior) {
+  return prior;
+}
+
+/**
+ * Inserts the given key-value pair into the map, and then recursively calls
+ * AddPairs to add any remaining arguments.
+ *
+ * @param prior A map into which the values should be inserted.
+ * @param key The key naming the field to insert.
+ * @param value A value to wrap with a call to Value(), above.
+ * @param rest Any remaining arguments
+ *
+ * @return The resulting map.
+ */
+template <typename ValueType, typename... Args>
+model::FieldValue::Map AddPairs(const model::FieldValue::Map& prior,
+                                const std::string& key,
+                                const ValueType& value,
+                                Args... rest) {
+  return AddPairs(prior.insert(key, Value(value)), rest...);
+}
+
+/**
+ * Creates an immutable sorted map from the given key/value pairs.
+ *
+ * @param key_value_pairs Alternating strings naming keys and values that can
+ *     be passed to Value().
+ */
+template <typename... Args>
+model::FieldValue::Map MakeMap(Args... key_value_pairs) {
+  return AddPairs(model::FieldValue::Map(), key_value_pairs...);
+}
+
+}  // namespace details
+
+/** Wraps an immutable sorted map in FieldValue with Type::Object. */
+inline model::FieldValue Object(const model::FieldValue::Map& value) {
```

Comment:
I think inheritance for the interface to `FieldValue` in C++ is problematic because `FieldValue` is used everywhere I'd prefer not to write that as `std::shared_ptr<FieldValue>`. If anything I think inheritance internally can make sense but then that makes parallel external inheritance weirder.

We could do `ObjectValue` extends `FieldValue` without pointers, but then we'd have to assert that they're the same size so as to avoid slicing. At that point it seems like a syntactic difference between what we have now and the inheritance way, except with added danger.

I still don't entirely understand what you mean by "a bit more natural". With is-a we cast. With has-a we have converters. They seem otherwise equivalent, though has-a has the added benefit that it resists slicing specifically because we're not casting.

Reply:
> "a bit more natural"

It's mostly that I was objecting to something named "object" (or "wrapobject") returning anything other than objectvalue. Looking at the current state of #3047, I've effectively renamed this to Map, returning a FV::Map type.

If we don't merge 3047 (or something like it) into this one, then... I'm not sure what to do here. The current state certainly isn't wrong... it just doesn't feel like as strong a guarantee as it could be. I suppose we could just merge as is.